Attachment support for crash reports

rqbacktrace opened the pull request:
This adds attachment support functionality for crash reports on `backtrace-cocoa`

I added this functionality to the demo app as well and manually tested live and crash reports and see the correct file attachment in both cases. I also manually tested live and crash reports without file attachments and they both still submitted correctly with no file attachments.

This does not fix a currently known bug where live-report specific attachments are not saved if that live-report is submitted on a retry. We can potentially use the functionality added here to fix that bug.

@KrzaQ I tagged you because I used your approach adding file attachment support to Crashpad, you might be better able to spot pitfalls in this approach

konraddysput commented on Sources/Public/Internal/SignalContext.swift:
```diff
@@ -2,7 +2,10 @@ import Foundation
+    // We use attachment paths as strings in our report send functions
+    var attachmentPathsArray: [ String ] { get }
     var attributes: Attributes { get set }
+    var attachments: Attachments { get set }
```
` our report send functions` <- report send function.. responsible for sending crashes? 
Can we also add description to `attachments` variable to avoid confussion later?

rqbacktrace replied:
I will clean this up, thank you!

konraddysput commented on Tests/AttachmentStorageTests.swift:
```diff
@@ -0,0 +1,60 @@
+        describe("AttachmentStorage") {
+            it("can save attachments as a plist") {
+                var crashAttachments = Attachments()
+                guard let fileUrl = try? self.createAFile() else {
```
Do we create physical file on hard drive?

rqbacktrace replied:
Yes, we are creating a physical file on the hard drive here.

konraddysput replied:
Are you sure we should create file in unit tests project? Unit tests shouldn't create files on the hard drive. @krml19 what do you think?

krml19 replied:
Yes, I agree. It'd be definitely better to use a mock

rqbacktrace replied:
@krml19 @konraddysput Even if I mock my dummy file here, `AttachmentsStorage.store` function would still write a plist containing the attachment paths, just like `AttributesStorage.store` function writes a plist containing attributes. Is it still worth it to create a mock filesystem here for my dummy file? We wouldn't fully eliminate file system writes/dependencies anyway because of `AttachmentsStorage.store` internal details.

krml19 replied:
I think it can be resolved by the dependency injection technique. In other words, one can swap and inject all the required external dependencies.

rqbacktrace replied:
@krml19 @konraddysput I have injected all filesystem dependencies, please take another look!

konraddysput reviewed (changes requested):
Hey - can we rethink attachmentStorage logic?

rqbacktrace commented on Examples/Example-iOS/AppDelegate.swift:
```diff
@@ -33,7 +47,23 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
         BacktraceClient.shared?.delegate = self
         BacktraceClient.shared?.attributes = ["foo": "bar", "testing": true]
-        
+        let fileName = "sample.txt"
```
You are right, I will make this a separate function to make it more clear for users how to use our `backtrace-cocoa` library

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,131 @@
+            do { let bookmark = try attachment.value.bookmarkData(options: URL.BookmarkCreationOptions.minimalBookmark)
+                attachmentsBookmarksDict[attachment.key] = bookmark as NSData
+            } catch {
+                print("Caught error: \(error)")
```
Ah thank you for catching this, I will clean this up!

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,49 @@ extension BacktraceCrashReporter: CrashReporting {
-        // NOTE: - no attachments in crash reports
-        return try BacktraceReport(report: reportData, attributes: attributes, attachmentPaths: [])
+        let attachmentPaths = copiedFileAttachments.map {$0.path}
+        return try BacktraceReport(report: reportData, attributes: attributes, attachmentPaths: attachmentPaths)
```
I don't think so. The crash report will contain a plist with its own attachment paths from the previous session. If we add attachments user declared in the current user session, we will potentially send double attachments, one from previous session and one from current user session. It could be a cool feature but maybe it can cause confusion

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,49 @@ extension BacktraceCrashReporter: CrashReporting {
+    // from pending crashes so that they are not overwritten by the
+    // new app session
+    func copyFileAttachmentsFromPendingCrashes() -> [URL] {
+        guard let cacheDirectoryUrl = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
```
That's a good idea, I will try it!

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,49 @@ extension BacktraceCrashReporter: CrashReporting {
+                try fileManager.copyItem(at: attachment.value, to: copiedAttachmentPath)
+                copiedFileAttachments.append(copiedAttachmentPath)
+            } catch {
+                print("Caught error: \(error)")
```
Ah thank you, will remove!

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,49 @@ extension BacktraceCrashReporter: CrashReporting {
+                if fileManager.fileExists(atPath: copiedAttachmentPath.path) {
+                    try fileManager.removeItem(atPath: copiedAttachmentPath.path)
+                }
+                try fileManager.copyItem(at: attachment.value, to: copiedAttachmentPath)
```
That's an awesome point, thank you Konrad!

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,131 @@
+        let directoryUrl: URL
+        let fileUrl: URL
+        
+        init(fileName: String) throws {
```
You're right...let me think how I can better unify this with our existing `AttributesStorage.swift`

rqbacktrace replied:
I think I have figured out a way to unify these using a common protocol. Please see the latest version!

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,49 @@ extension BacktraceCrashReporter: CrashReporting {
+        let attachments = (try? AttachmentsStorage.retrieve(fileName: BacktraceCrashReporter.crashName)) ?? [:]
+        var copiedFileAttachments = [URL]()
+        for attachment in attachments {
+            let fileManager = FileManager()
```
I don't think it will work, cacheDirectoryUrl calls a static function of FileManager (`FileManager.default.urls`). Otherwise for our file operations below we need an actual FileManager instance.

konraddysput replied:
`FileManager.default` is your Filemanager instance.

rqbacktrace replied:
We don't use `FileManager.default` here anymore. I agree with your feedback that it's better to get this path directly from `AttachmentsStorage`.

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,131 @@
+/// Expected format: Filename, File URL bookmark
+public typealias Bookmarks = [String: NSData]
+
+final class AttachmentsStorage {
```
It won't be easy to unify these since the function parameters are a bit different. Even the Configs are different (`AttachmentsStorage.swift` needs a different plist path than `AttributesStorage.swift`). Even so, I think it can be done, but it will involve significant refactoring. It's probably a good idea for us to do this, especially if we want to add other kinds of `Storage` classes for other kinds of report metadata.

rqbacktrace replied:
I think it looks much better now. Thank you for this good suggestion! Please see the updated version!

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+        
+        guard let bookmarks = dictionary as? Bookmarks else {
+            BacktraceLogger.debug("Could not convert stored dictionary to Bookmarks type")
+            throw FileError.unsupportedScheme
```
The thrown error does not match the failure reason.

rqbacktrace replied:
I will create a new error type for these errors, thank you!

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+        }
+        guard let attachments = try? extractAttachmentUrls(bookmarks) else {
+            BacktraceLogger.debug("Could not extract attachment URLs from stored attachments Bookmarks")
+            throw FileError.invalidPropertyList
```
The thrown error does not match the failure reason.

rqbacktrace replied:
I will create a new error type for these errors, thank you!

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+        guard let attachments = try? retrieve(fileName: fileName) else {
+            return []
+        }
+        var attachmentsArray = [ String ]()
+        for attachment in attachments {
+            attachmentsArray.append(attachment.value.path)
+        }
+        return attachmentsArray
```
Saw it in other places as well. You might want to think of adding a method or a computed property to get this value 

```suggestion
let attachmentsArray = attachments.map(\.value.path)
```

rqbacktrace replied:
I think I can remove this function, it was only used in tests anyway. Thank you for showing me how to use `map` function, this is very helpful for writing clean Swift code.

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+    private static func convertAttachmentUrlsToBookmarks(_ attachments: Attachments) throws -> Bookmarks {
+        var attachmentsBookmarksDict = Bookmarks()
+        for attachment in attachments {
+            do { let bookmark = try attachment.value.bookmarkData(options: URL.BookmarkCreationOptions.minimalBookmark)
```
Expression should start from the new line

rqbacktrace replied:
Will fix, thank you!

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+                continue
+            }
+            if stale.boolValue {
+                BacktraceLogger.error("Bookmark data is stale. This should not happen")
```
The error seems to be silently ignored. Is it intentional?

rqbacktrace replied:
We should handle this error since we're not supposed to use stale URLs. Thank you for pointing this out!

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+        var attachments = Attachments()
+        for bookmark in bookmarks {
+            var stale = ObjCBool(false)
+            guard let fileUrl = try? NSURL.init(resolvingBookmarkData: bookmark.value as Data,
```
```suggestion
            guard let fileUrl = try? NSURL(resolvingBookmarkData: bookmark.value as Data,
```
Can we replace `NSURL` with `URL`?

rqbacktrace replied:
Yes we can, thank you!

krml19 reviewed (changes requested):
Please adjust the style

rqbacktrace commented on Sources/Features/Attributes/AttributesProvider.swift:
```diff
@@ -43,6 +44,14 @@ extension AttributesProvider: SignalContext {
     }
     
+    var attachmentPathsArray: [ String ] {
+        var attachmentPathsArray = [ String ]()
+        for attachment in attachments {
+            attachmentPathsArray.append(attachment.value.path)
+        }
+        return attachmentPathsArray
+    }
```
Will do, thank you!

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+
+/// Type-alias of storing file attachments on disk (as a bookmark)
+/// Expected format: Filename, File URL bookmark
+public typealias Bookmarks = [String: NSData]
```
Will make these changes, thank you!

rqbacktrace commented on Sources/Public/Internal/SignalContext.swift:
```diff
@@ -3,6 +3,10 @@ import Foundation
+    // File attachments are stored to disk as URLs
+    var attachments: Attachments { get set }
+    // File attachments are used in BacktraceReports as string paths
+    var attachmentPathsArray: [ String ] { get }
```
Will do, thank you!

rqbacktrace commented on Sources/Public/Internal/SignalContext.swift:
```diff
@@ -3,6 +3,10 @@ import Foundation
     var attributes: Attributes { get set }
+    // File attachments are stored to disk as URLs
+    var attachments: Attachments { get set }
+    // File attachments are used in BacktraceReports as string paths
```
Will do, thank you!

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+        guard let attachments = try? retrieve(fileName: fileName) else {
+            return []
+        }
+        var attachmentsArray = [ String ]()
```
Will do, thank you!

rqbacktrace commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,129 @@
+/// Expected format: Filename, File URL bookmark
+public typealias Bookmarks = [String: NSData]
+
+final class AttachmentsStorage {
```
Will do, thank you! I will make this change to `AttributesStorage.swift` as well!

krml19 commented on Sources/Public/Internal/SignalContext.swift:
```diff
@@ -3,6 +3,10 @@ import Foundation
 protocol SignalContext: CustomStringConvertible {
     var allAttributes: Attributes { get }
     var attributes: Attributes { get set }
+    // File attachments are stored to disk as URLs
+    var attachments: Attachments { get set }
+    // File attachments are used in `BacktraceReport` as string paths
+    var attachmentPathsArray: [String] { get }
```
What's the reason for storing `attachments` and `attachmentPathsArray`? Also, I suggest renaming `attachementPathsArray` to `attachmentPaths`.

rqbacktrace replied:
We need `attachmentPathsArray` for live reports. I tagged you where we use `attachmentPathsArray` in `BacktraceReporter`. 

`attachmentPaths` is definitely a better name, will rename.

krml19 commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -52,19 +56,62 @@ extension BacktraceCrashReporter: CrashReporting {
     }
+    
+    func deleteCopiedFileAttachments() throws {
+        let fileManager = FileManager()
```
```suggestion
        let fileManager = FileManager.default
```

krml19 commented on Sources/Features/Client/Model/AttachmentsStorage.swift:
```diff
@@ -0,0 +1,125 @@
+        var attachments = Attachments()
+        for bookmark in bookmarks {
+            var stale = Bool(false)
+            guard let fileUrl = try? URL(resolvingBookmarkData: bookmark.value,
+                                                options: URL.BookmarkResolutionOptions(),
+                                                relativeTo: nil,
+                                                bookmarkDataIsStale: &stale) else {
```
Please align all parameters

rqbacktrace commented on Sources/Features/Client/BacktraceReporter.swift:
```diff
@@ -96,7 +104,7 @@ extension BacktraceReporter {
         let resource = try reporter.generateLiveReport(exception: exception,
                                                        attributes: attributesProvider.allAttributes,
-                                                       attachmentPaths: attachmentPaths)
+                                                       attachmentPaths: attachmentPaths + attributesProvider.attachmentPathsArray)
```
@krml19 this is why we need `attachmentPathsArray` parameter. But I will change it to `attachmentPaths`, that is definitely a better name

rqbacktrace commented:
@krml19 @konraddysput please see latest version!

rqbacktrace commented:
@krml19 I have refactored `AttachmentsStorage` and `AttributesStorage` to allow injection of `ReportMetadataStorage` and updated the unit tests to use a mocked `ReportMetadataStorageMock`. Please have another look! Thank you!

rqbacktrace commented on Sources/Public/BacktraceCrashReporter.swift:
```diff
@@ -5,6 +5,7 @@ import Backtrace_PLCrashReporter
 @objc public class BacktraceCrashReporter: NSObject {
     private let reporter: PLCrashReporter
     static private let crashName = "live_report"
+    private var copiedFileAttachments = [URL]()
```
Done, thank you!

konraddysput reviewed (approved):
Nothing critical on my side. Thanks a lot @rqbacktrace !

rqbacktrace commented on Examples/Example-iOS/AppDelegate.swift:
```diff
@@ -46,6 +52,30 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
+        do {
+            try myData.write(to: fileUrl, atomically: true, encoding: .utf8)
+        } catch {
+            print("Error: \(error)")
```
@konraddysput apologies...I don't quite understand. Can you please give me some more details?

konraddysput replied:
Right now we're printing error to stdout. Does it make sense to fail instead? Because otherwise user will assume attachment support doesn't work.

konraddysput replied:
What do you think?

rqbacktrace replied:
Ah good idea @konraddysput, I will make this change

krml19 commented on Sources/Public/BacktraceClientCustomizing.swift:
```diff
@@ -6,12 +6,23 @@ public typealias BacktraceClientProtocol = BacktraceReporting & BacktraceClientC
 
+/// Type-alias of passing file attachments to library.
+/// Expected format: Filename, File URL bookmark
+public typealias Attachments = [String: URL]
+
+/// Type-alias of storing file attachments on disk (as a bookmark)
+/// Expected format: Filename, File URL bookmark
+public typealias Bookmarks = [String: Data]
```
Does it need to be public?

rqbacktrace replied:
@krml19 if it's private I won't be able to use it in my `AttachmentBookmarkHandlerMock` class

krml19 reviewed (approved):
LGTM 👍
